### PR TITLE
[BREAKING] Allocate local variables according to their scope

### LINKED ISSUE #3222: Feature request: mechanism for removing variables from the stack
I'd like to propose adding a mechanism to Solidity that allows removing variables from the stack.  I'm not familiar with Solidity's internals, so I do not have an opinion about how to implement this feature.  However, I assume it is _possible_ to implement because Serpent had this feature.  For reference, Serpent's `with` keyword was used for this:
```
# nothing on the stack
with x = 1:
    # x is on the stack
# nothing on the stack
```

**Motivation:** Augur recently completed its migration to Solidity from Serpent.  One pain point for us was that Solidity lacks a way of removing variables from the stack.  Because of this, the stack depth limit drove a number of architectural decisions (about e.g. how/whether to split up functions) that we would have preferred to make independently of language-level constraints.

(This is related to https://github.com/ethereum/solidity/issues/2693, but I thought it was worthwhile to make a separate issue for manual stack management.)

### PR DESCRIPTION
**Pull request by leonardoalt**
Implements #3672. Closes #3222.

There are two scenarios so far:

1) No `break`/`continue`. That's the simple case, local vars are allocated when declared, and freed when the execution leaves their scope.

2) `Break`. Every `break` statement inside a loop requires a different amount of `pop`s, depending on the amount of local variables declared until that point.
```
function f(bool b, bool c, bool d) public pure {
	uint x;
	for (uint i = 0; i < 10; ++i) {
		uint y = 12;
		if (b) break;
		if (c) {
			uint z = 39;
			if (d) break;
			x += z;
		}
		x += y;
	}
}
```
In this example, the second `break` jumps to `tag_14`, since it requires 3 pops: `z`, `y` and `i`.
The first `break` jumps to `tag_11`, since it needs to free only `y` and `i`.
Any execution where a `break` happens, after freeing the local vars, jumps to the actual end of the loop, in this case `tag_15`.
Here we have `tag_8` as the end of the loop for executions that do not have any `break` (loop condition failure jumps here), in which case we know that the local variables are freed in the end of their scopes (as in `1`) and we only have to free loop init vars, in this case `i`.
```
   ...
    tag_14:
      pop
    tag_11:
      pop
      pop
      jump(tag_15)
    tag_8:
        /* "break.sol":135:141  uint i */
      pop
        /* "break.sol":130:273  for (uint i = 0; i < 10; ++i) {... */
    tag_15:
    ...
```

I had to comment out two stack height checks (to be adjusted manually) and two other checks such that exceptions do not occur:
libevmasm/Assembly.h:94
libevmasm/Assembly.cpp:328

I'm investigating the exceptions.

### CODE REVIEW

**Comment by leonardoalt** [comment]
The reason for the `stack underflow` is basically the same: jumps are not taken into account in the stack height.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 1]
```diff
@@ -691,7 +687,13 @@ bool ContractCompiler::visit(WhileStatement const& _whileStatement)
 		m_context.appendConditionalJumpTo(loopEnd);
 	}
 
+	solAssert(m_loops.top() == &_whileStatement, "");
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: What if `m_loops.empty()`? This and the next line would also be a good candidate for extracting into a function (I assume it is also used by other loops).

File: libsolidity/codegen/ContractCompiler.cpp [hunk 2]
```diff
@@ -691,7 +687,13 @@ bool ContractCompiler::visit(WhileStatement const& _whileStatement)
...
+	m_loops.pop();
+
 	m_context.appendJumpTo(loopStart);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: I kind of have the impression that this should move above the `solAssert`, although that is subjective.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Then it should definitely break :p 
Right, makes sense

File: libsolidity/codegen/ContractCompiler.cpp [hunk 3]
```diff
@@ -979,3 +1018,56 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
...
+		m_loopScopedVariables[m_loops.top()].push_back(&_decl);
+	appendStackVariableInitialisation(_decl);
+}
+
+void ContractCompiler::popBlockScopedVariables(ASTNode const* _node)
+{
+	for (auto _decl : m_scopedVariables[_node])
+	{
+		CompilerContext::LocationSetter locationSetter(m_context, *_decl);
+		m_context << Instruction::POP;
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: This should be `CompilerUtils utils(m_context).popStackElement(_decl->type())` - not all types occupy exactly one stack slot.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 4]
```diff
@@ -979,3 +1018,56 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
...
+	}
+	m_scopedVariables.erase(_node);
+}
+
+void ContractCompiler::popLoopScopedVariables(ASTNode const* _node)
+{
+	for (auto _decl : m_loopScopedVariables[_node])
+	{
+		CompilerContext::LocationSetter locationSetter(m_context, *_decl);
+		m_context << Instruction::POP;
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Same here.

**Comment by leonardoalt** [comment]
Missing:

- [X] Make sure `return` is working
- [x] Document properly
- [x] Modifiers

File: libsolidity/codegen/ContractCompiler.cpp [hunk 5]
```diff
@@ -495,16 +494,20 @@ bool ContractCompiler::visit(FunctionDefinition const& _function)
...
 
+
 	for (ASTPointer<VariableDeclaration const> const& variable: _function.parameters() + _function.returnParameters())
 		m_context.removeVariable(*variable);
-	for (VariableDeclaration const* localVariable: _function.localVariables())
-		m_context.removeVariable(*localVariable);
 
 	m_context.adjustStackOffset(-(int)c_returnValuesSize);
 
+	CompilerUtils(m_context).popStackSlots(stackSurplus);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Why is this needed? Why can't this be done right after calling `appendModifierOrFunctionCode();`?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 6]
```diff
@@ -733,28 +736,59 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
...
+	solAssert(!m_loops.empty(), "");
+	unsigned sizeOnStack = 0;
+	for (auto _decl : m_loopScopedVariables[m_loops.back()])
+		sizeOnStack += _decl->type()->sizeOnStack();
+	m_breakTagsPops.push_back(std::make_pair<unsigned, eth::AssemblyItem>(
+		unsigned(sizeOnStack),
+		m_context.newTag()
+	));
+	if (dynamic_cast<Break const*>(_statement))
+		m_context << 0;
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Whit is the purpose of these constants?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 7]
```diff
@@ -780,10 +814,16 @@ bool ContractCompiler::visit(Return const& _return)
...
-	for (unsigned i = 0; i < m_stackCleanupForReturn; ++i)
-		m_context << Instruction::POP;
-	m_context.appendJumpTo(m_returnTags.back());
-	m_context.adjustStackOffset(m_stackCleanupForReturn);
+
+	unsigned sizeOnStack = 0;
+	for (auto _varDecl : m_scopedVariables)
+		for (auto _decl : _varDecl.second)
+			sizeOnStack += _decl->type()->sizeOnStack();
+	m_returnTagsPops.push_back(std::make_pair<unsigned, eth::AssemblyItem>(
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Why don't you just remove the local variables in the same way it was done before?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 8]
```diff
@@ -912,34 +964,22 @@ void ContractCompiler::appendModifierOrFunctionCode()
...
 	{
-		m_returnTags.push_back(m_context.newTag());
-
 		codeBlock->accept(*this);
 
-		solAssert(!m_returnTags.empty(), "");
-		m_context << m_returnTags.back();
-		m_returnTags.pop_back();
-
-		CompilerUtils(m_context).popStackSlots(stackSurplus);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Why not just keep this?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 9]
```diff
@@ -979,3 +1019,129 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
...
+	}
+	m_scopedVariables.erase(_node);
+}
+
+/**
+Whenever a break or continue is reached, it counts the number of
+stack slots that should be freed at that moment with respect to
+local variables, and creates a new tag that leads to the freeing
+of those variables.
+All the break/continue statements inside the same loop share this
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: While this is a clever technique to save some code, this is disastrous for bytecode analysis in the optimizer and other tools. Also I don't think it happens too often that we have multiple break / continue statements in a loop where this would be beneficial.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 7 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: To use the optimized shared structure thingy.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 6 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Allowing breaks and continues to share the pop structure. In the end of the pops, a 0 would lead to a jump to outside the loop and a 1 would jump back to the loop condition test.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 9 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Ok ;)
Confirming then: each break/continue/return should point to their own non-shared structure that jumps accordingly at the end.

**Comment by leonardoalt** [comment]
The reason why I removed the `returnTags` from after the visit of the function/modifier block is that the local variables from that block will be freed, but this shouldn't get mixed with the parameters being freed which is also necessary after the pops for local variables that a `return` does.
That's why the stack on size of the modifier parameters are summed and only freed when the function parameters are also freed.

**Comment by chriseth** [comment]
@leonardoalt can you comment on what you want to do with the code block that is jumped to for break and continue?

**Comment by leonardoalt** [comment]
The idea now is
- `break`: each break pops the local variables that were declared inside the latest loop and were not freed yet, and jumps to outside the loop. In the case of a `for` loop, the init var is **not** freed, that is done by the `forStatement` itself.
- `continue`: each continue pops the local variables that were declared inside the latest loop and were not freed yet, and jumps to the loop condition test.
- `return`: each return pops all the local variables that were declared inside the function **and** modifiers and were not freed yet, and jumps to the end of the function where the parameters of the function and modifiers are freed. This caused `m_returnTags` to contain only one tag (to be cleaned up) which is used by all `return`s. Another consequence of this is the following change: before, modifiers local vars and parameters were freed together. Now the local variables are freed either in the normal execution when the `Block` is `endVisit`ed or in a `return` statement, and the parameters are freed together with the other function and modifier parameters in the end. The reason is that the `return` pop blocks do not free the parameters and need to jump to that point.

**Comment by chriseth** [comment]
Rebased.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 10]
```diff
@@ -737,28 +746,62 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
...
+}
+
+bool ContractCompiler::visitBreakContinue(Statement const* _statement)
+{
+	CompilerContext::LocationSetter locationSetter(m_context, *_statement);
+	unsigned sizeOnStack = loopVariablesSize();
+	// The context stack needs to be adjusted because the following
+	// pops are extra in the sense that they don't come from a normal
+	// block endVisit.
+	m_context.adjustStackOffset(sizeOnStack);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: It is probably better to move this to the end of the function, because the adjustment has to be made for the dangling block that comes after the unconditional jump.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 11]
```diff
@@ -737,28 +746,62 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
...
-	CompilerContext::LocationSetter locationSetter(m_context, _breakStatement);
-	if (!m_breakTags.empty())
+	return visitBreakContinue(&_breakStatement);
+}
+
+unsigned ContractCompiler::loopVariablesSize()
+{
+	solAssert(!m_loops.empty(), "");
+	unsigned sizeOnStack = 0;
+	for (auto _decl : m_loopScopedVariables[m_loops.back()])
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: no space in front of `:`

File: libsolidity/codegen/ContractCompiler.cpp [hunk 12]
```diff
@@ -737,28 +746,62 @@ bool ContractCompiler::visit(ForStatement const& _forStatement)
...
 }
 
 bool ContractCompiler::visit(Break const& _breakStatement)
 {
-	CompilerContext::LocationSetter locationSetter(m_context, _breakStatement);
-	if (!m_breakTags.empty())
+	return visitBreakContinue(&_breakStatement);
+}
+
+unsigned ContractCompiler::loopVariablesSize()
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Naming: what about `stackSizeOfCurrentLoopVariables`?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 13]
```diff
@@ -784,10 +827,15 @@ bool ContractCompiler::visit(Return const& _return)
 		for (auto const& retVariable: boost::adaptors::reverse(returnParameters))
 			CompilerUtils(m_context).moveToStackVariable(*retVariable);
 	}
-	for (unsigned i = 0; i < m_stackCleanupForReturn; ++i)
-		m_context << Instruction::POP;
+
+	unsigned sizeOnStack = 0;
+	for (auto _varDecl : m_scopedVariables)
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: No space in front of `:`

File: libsolidity/codegen/ContractCompiler.cpp [hunk 14]
```diff
@@ -784,10 +827,15 @@ bool ContractCompiler::visit(Return const& _return)
 		for (auto const& retVariable: boost::adaptors::reverse(returnParameters))
 			CompilerUtils(m_context).moveToStackVariable(*retVariable);
 	}
-	for (unsigned i = 0; i < m_stackCleanupForReturn; ++i)
-		m_context << Instruction::POP;
+
+	unsigned sizeOnStack = 0;
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: It might make sense to create a helper function that does one of the following:

a) in input of a number and a tag, pops this number of stack slots, appends a jump to the tag and then adjusts the stack offset

b) on input of a set or vector of variable declarations, counts the stack slots and does as in a)

File: libsolidity/codegen/ContractCompiler.cpp [hunk 15]
```diff
@@ -810,8 +858,14 @@ bool ContractCompiler::visit(EmitStatement const& _emit)
 
 bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclarationStatement)
 {
-	StackHeightChecker checker(m_context);
 	CompilerContext::LocationSetter locationSetter(m_context, _variableDeclarationStatement);
+
+	// Local variable slots are reserved when their declaration is seen,
+	// and freed in the end of their scope.
+	for (auto _decl : _variableDeclarationStatement.declarations())
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: No space in front of `:`

File: libsolidity/codegen/ContractCompiler.cpp (hunk 8 above)
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: I still don't understand why we do not have to pop the modifier parameters here.

Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Because this is also where the `endVisit` of the modifier block frees the modifier local variables in a normal execution (without return). Since a `return` will also pop the modifier local variables, it has to skip this part. Then all the execution paths converge to parameter freeing which happens at the very end.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 12 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Yea, I agree the naming isn't very precise, I just usually run away from long names :p

File: libsolidity/codegen/ContractCompiler.h [hunk 16]
```diff
@@ -132,7 +132,14 @@ class ContractCompiler: private ASTConstVisitor
...
 	/// Calculates the amount of stack slots allocated for
 	/// local variables inside the latest loop.
-	unsigned loopVariablesSize();
+	unsigned stackSizeOfCurrentLoopVariables();
+
+	/// Calculates the amount of stack slots allocated for
+	/// local variables currently allocated.
+	unsigned stackSizeOfCurrentLocalVariables();
+
+	/// Pops _amount slots from the stack and jumps to _jumpTo
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.h]: And readjusts the stack offset to the original value.

**Comment by chriseth** [comment]
Please rebase and reset PR target branch.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 17]
```diff
@@ -495,16 +491,16 @@ bool ContractCompiler::visit(FunctionDefinition const& _function)
 		}
 	//@todo assert that everything is in place now
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Perhaps we should actually implement this now.
This would be a check that the elements in `stackLayout` are `0,1,2,....,k`

File: libsolidity/codegen/ContractCompiler.cpp [hunk 18]
```diff
@@ -810,8 +828,15 @@ bool ContractCompiler::visit(EmitStatement const& _emit)
 
 bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclarationStatement)
 {
-	StackHeightChecker checker(m_context);
 	CompilerContext::LocationSetter locationSetter(m_context, _variableDeclarationStatement);
+
+	// Local variable slots are reserved when their declaration is seen,
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: `visited` instead of `seen`? (seen could refer to "being in the same scope" or something like that)

File: libsolidity/codegen/ContractCompiler.cpp [hunk 19]
```diff
@@ -983,3 +1006,68 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
 	a << u256(0x20) << u256(0) << Instruction::RETURN;
 	return make_shared<eth::Assembly>(a);
 }
+
+void ContractCompiler::addScopedVariable(VariableDeclaration const& _decl)
+{
+	m_scopedVariables[m_modifierDepth][_decl.scope()].push_back(&_decl);
+	if (stackSizeOfCurrentLocalVariables() > 17)
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Do we really need this here? It might be that a variable is added that is never accessed.

Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: Or actually something like this did not compile in the old version:

```
{
 unit a1; 
 a1 = 7;
 uint a2; ... uint a15; uint a16;
 { uint a17 = f(); a16 = a17; }
}
```
but it will with in the current version.
```

File: libsolidity/codegen/ContractCompiler.cpp [hunk 20]
```diff
@@ -983,3 +1006,68 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
...
+unsigned ContractCompiler::stackSizeOfCurrentLocalVariables()
+{
+	unsigned sizeOnStack = 0;
+	for (auto _varDecl: m_scopedVariables[m_modifierDepth])
+		for (auto _decl: _varDecl.second)
+			sizeOnStack += _decl->type()->sizeOnStack();
+	return sizeOnStack;
+}
+
+void ContractCompiler::popAndJump(unsigned _amount, eth::AssemblyItem const& _jumpTo)
```
Comment (axic) [inline libsolidity/codegen/ContractCompiler.cpp]: Is it worth splitting this and merging now?

File: libsolidity/codegen/ContractCompiler.cpp (hunk 19 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: That compiles, but adding another unused variable breaks compilation.
If we remove the `if` you mentioned, where is this error caught later on?

Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: The error should be caught at the point where the value of the variable is requested or a new value is to be stored in the variable. Note that the limit is actually smaller inside more complex expressions, so this check here would not suffice anyway.

Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Ok, gonna remove it

**Comment by leonardoalt** [comment]
Fixed suggestions, rebased and squashed.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 21]
```diff
@@ -810,8 +829,15 @@ bool ContractCompiler::visit(EmitStatement const& _emit)
...
 bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclarationStatement)
 {
-	StackHeightChecker checker(m_context);
 	CompilerContext::LocationSetter locationSetter(m_context, _variableDeclarationStatement);
+
+	// Local variable slots are reserved when their declaration is visited,
+	// and freed in the end of their scope.
+	for (auto _decl: _variableDeclarationStatement.declarations())
+		if (_decl)
+			addScopedVariable(*_decl);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: This currently always zero-initializes variables, even if they are assigned to in the same statement, we might optimize that in a future PR.

Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Yes, I'll create an issue an assign to myself to keep track of it.

**Comment by leonardoalt** [comment]
@chriseth Simplified/reimplemented some parts using the break/continue/return tags with the stack height.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 22]
```diff
@@ -441,22 +441,18 @@ bool ContractCompiler::visit(FunctionDefinition const& _function)
...
 	for (ASTPointer<VariableDeclaration const> const& variable: _function.returnParameters())
 		appendStackVariableInitialisation(*variable);
-	for (VariableDeclaration const* localVariable: _function.localVariables())
-		appendStackVariableInitialisation(*localVariable);
 
 	if (_function.isConstructor())
 		if (auto c = m_context.nextConstructor(dynamic_cast<ContractDefinition const&>(*_function.scope())))
 			appendBaseConstructor(*c);
 
-	solAssert(m_returnTags.empty(), "");
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Why did you have to remove this?

File: libsolidity/codegen/ContractCompiler.h [hunk 23]
```diff
@@ -123,19 +125,35 @@ class ContractCompiler: private ASTConstVisitor
 	/// @returns the runtime assembly for clone contracts.
 	eth::AssemblyPointer cloneRuntime() const;
 
+	/// Frees the variables of a certain scope (to be used when leaving).
+	void popScopedVariables(ASTNode const* _node);
+
+	/// Pops _amount slots from the stack and jumps to _jumpTo.
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.h]: I usually use `@a _amount` - but actually I don't know what that means...

File: libsolidity/codegen/ContractCompiler.h [hunk 24]
```diff
@@ -123,19 +125,35 @@ class ContractCompiler: private ASTConstVisitor
...
+	/// Also readjusts the stack offset to the original value.
+	void popAndJump(unsigned _amount, eth::AssemblyItem const& _jumpTo);
+
+	/// Sets the stack height for the visited loop.
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.h]: From the name I would say that the loop is also visited. Perhaps rename to `loopVisited` or `storeLoopStackHeight`?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 25]
```diff
@@ -983,3 +996,26 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
...
+	if (m_scopeStackHeight[m_modifierDepth].size() == 0)
+		m_scopeStackHeight.erase(m_modifierDepth);
+}
+
+void ContractCompiler::visitLoop(BreakableStatement const* _loop)
+{
+	m_scopeStackHeight[m_modifierDepth][_loop] = m_context.stackHeight();
+}
+
+void ContractCompiler::popAndJump(unsigned _amount, eth::AssemblyItem const& _jumpTo)
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: This might be better suited inside the context itself.

File: libsolidity/codegen/ContractCompiler.h [hunk 26]
```diff
@@ -123,19 +125,35 @@ class ContractCompiler: private ASTConstVisitor
...
-	std::vector<eth::AssemblyItem> m_breakTags; ///< tag to jump to for a "break" statement
-	std::vector<eth::AssemblyItem> m_continueTags; ///< tag to jump to for a "continue" statement
-	/// Tag to jump to for a "return" statement, needs to be stacked because of modifiers.
-	std::vector<eth::AssemblyItem> m_returnTags;
+	/// Tag to jump to for a "break" statement and the stack height after freeing the local loop variables.
+	std::vector<std::pair<eth::AssemblyItem, unsigned>> m_breakTags;
+	/// Tag to jump to for a "continue" statement and the stack height after freeing the local loop variables.
+	std::vector<std::pair<eth::AssemblyItem, unsigned>> m_continueTags;
+	/// Tag to jump to for a "return" statement and the stack height after freeing the local function or modifier variables.
+	/// Needs to be stacked because of modifiers.
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.h:143]: I think in a future refactor, we should collect all these fields that depend on the modifier depth and put them in a single struct or even instantiate a new compiler class.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 27]
```diff
@@ -983,3 +996,26 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
 	a << u256(0x20) << u256(0) << Instruction::RETURN;
 	return make_shared<eth::Assembly>(a);
 }
+
+void ContractCompiler::popScopedVariables(ASTNode const* _node)
+{
+	unsigned blockHeight = m_scopeStackHeight[m_modifierDepth][_node];
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Please assert that `_node` is a valid key or use `at()`.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 28]
```diff
@@ -784,10 +788,8 @@ bool ContractCompiler::visit(Return const& _return)
 		for (auto const& retVariable: boost::adaptors::reverse(returnParameters))
 			CompilerUtils(m_context).moveToStackVariable(*retVariable);
 	}
-	for (unsigned i = 0; i < m_stackCleanupForReturn; ++i)
-		m_context << Instruction::POP;
-	m_context.appendJumpTo(m_returnTags.back());
-	m_context.adjustStackOffset(m_stackCleanupForReturn);
+
+	popAndJump(m_context.stackHeight() - m_returnTags.back().second, m_returnTags.back().first);
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: It might be less error prone to only supply `m_returnTags.back().second` to `popAndJump` and do `m_context.stackHeight() - ...` inside `popAndJump`.

**Review by chriseth — CHANGES_REQUESTED** [review]
Please add some tests: for example for break, continue, return inside for and inside do-while (and while-do if needed).

File: libsolidity/codegen/ContractCompiler.cpp (hunk 25 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: What about inside `CompilerUtils` instead of the context?

File: libsolidity/codegen/ContractCompiler.cpp (hunk 22 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.cpp]: Forgot to put back.

File: libsolidity/codegen/ContractCompiler.h (hunk 24 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.h]: Reminded me that actually that's not needed, only for loops need to

File: libsolidity/codegen/ContractCompiler.cpp (hunk 25 above)
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: Also fine!

File: libsolidity/codegen/ContractCompiler.h (hunk 26 above)
Comment (leonardoalt) [reply libsolidity/codegen/ContractCompiler.h:143]: Makes sense, I'll create an issue to remember that.

File: libsolidity/codegen/CompilerContext.cpp [hunk 29]
```diff
@@ -130,14 +130,27 @@ void CompilerContext::addVariable(VariableDeclaration const& _declaration,
...
 		m_localVariables.erase(&_declaration);
 }
 
+void CompilerContext::removeVariablesAboveStackHeight(unsigned _stackHeight)
+{
+	vector<Declaration const*> toRemove;
+	for (auto _var: m_localVariables)
+	{
+		solAssert(!_var.second.empty(), "");
+		if (_var.second.back() >= _stackHeight)
```
Comment (chriseth) [inline libsolidity/codegen/CompilerContext.cpp]: Might it make sense to add the following two safety measures?
 - assert here that `_var.second.back()` is at most the current stack height
 - assert that `m_localVariables` is empty at the end of the function

File: libsolidity/codegen/CompilerUtils.cpp [hunk 30]
```diff
@@ -1170,6 +1170,14 @@ void CompilerUtils::popStackSlots(size_t _amount)
 		m_context << Instruction::POP;
 }
 
+void CompilerUtils::popAndJump(unsigned _toHeight, eth::AssemblyItem const& _jumpTo)
+{
+	unsigned amount = m_context.stackHeight() - _toHeight;
```
Comment (chriseth) [inline libsolidity/codegen/CompilerUtils.cpp]: Please either add an assertion that we do not get an overflow here or change the type to `int`.

File: libsolidity/codegen/ContractCompiler.cpp [hunk 31]
```diff
@@ -983,3 +995,19 @@ eth::AssemblyPointer ContractCompiler::cloneRuntime() const
 	a << u256(0x20) << u256(0) << Instruction::RETURN;
 	return make_shared<eth::Assembly>(a);
 }
+
+void ContractCompiler::popScopedVariables(ASTNode const* _node)
+{
+	unsigned blockHeight = m_scopeStackHeight.at(m_modifierDepth).at(_node);
+	unsigned stackDiff = m_context.stackHeight() - blockHeight;
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp]: Also add assertion here.

**Review by chriseth — CHANGES_REQUESTED** [review]
Sorry that I have to be the pedantic again, but it would be good to also have tests for:
 - nested for loops
 - variables declared in the same block as break / continue but after those statements
 - modifiers in conjunction with return (those migth already exist), break and continue
 - variables of stack size different than one

Examples of the latter are: calldata arrays (2), external function pointers (2) and actually I'm not sure if it is still possible to declare variables with size zero.

File: libsolidity/codegen/CompilerContext.cpp (hunk 29 above)
Comment (leonardoalt) [reply libsolidity/codegen/CompilerContext.cpp]: `m_localVariables` doesn't necessarily have to be empty in the end of the function, since it's called every time a scope closes

Comment (leonardoalt) [reply libsolidity/codegen/CompilerContext.cpp]: Or did you mean in the end of visit to `FunctionDefinition` in `ContractCompiler`?

**Comment by leonardoalt** [comment]
Added more tests for: nested for loops, local variables declared after break/continue, break/continue/return in modifiers with loop (there are also more complex already existing modifier tests), calldata array with break/continue and extra local variables.

File: libsolidity/codegen/CompilerContext.cpp [hunk 32]
```diff
@@ -127,6 +127,8 @@ void CompilerContext::addVariable(VariableDeclaration const& _declaration,
 								  unsigned _offsetToCurrent)
 {
 	solAssert(m_asm->deposit() >= 0 && unsigned(m_asm->deposit()) >= _offsetToCurrent, "");
+	unsigned sizeOnStack = _declaration.annotation().type->sizeOnStack();
+	solAssert(sizeOnStack == 1 || sizeOnStack == 2, "");
```
Comment (chriseth) [inline libsolidity/codegen/CompilerContext.cpp]: Perhaps add a comment: variables should not have other stack sizes, but that might change when new types are introduced.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 33]
```diff
@@ -9420,6 +9528,50 @@ BOOST_AUTO_TEST_CASE(break_in_modifier)
...
+		contract C {
+			uint public x;
+			modifier run() {
+				for (uint i = 0; i < 10; i++) {
+					if (i % 2 == 1) continue;
+					_;
+				}
+			}
+			function f() run {
+				x++;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please add a local variable here.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 34]
```diff
@@ -9420,6 +9528,50 @@ BOOST_AUTO_TEST_CASE(break_in_modifier)
...
+		contract C {
+			uint public x;
+			modifier run() {
+				for (uint i = 1; i < 10; i++) {
+					if (i == 5) return;
+					_;
+				}
+			}
+			function f() run {
+				x++;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: same here.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 35]
```diff
@@ -727,6 +775,66 @@ BOOST_AUTO_TEST_CASE(for_loop_multiple_local_vars)
...
+							uint t = z * 3;
+						}
+						x++;
+						if (x > 20) {
+							return 84;
+							uint h = x + 42;
+						}
+					}
+					if (x > 30) {
+					    return 42;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: indentation.

File: libsolidity/codegen/CompilerContext.cpp (hunk 29 above)
Comment (leonardoalt) [reply libsolidity/codegen/CompilerContext.cpp]: There's a test breaking the first assertion suggestion, debugging it.

Comment (leonardoalt) [reply libsolidity/codegen/CompilerContext.cpp]: Fixed

**Review by chriseth — APPROVED** [review]

File: libsolidity/codegen/ContractCompiler.cpp [hunk 36]
```diff
@@ -498,8 +498,12 @@ bool ContractCompiler::visit(FunctionDefinition const& _function)
 	m_context.adjustStackOffset(-(int)c_returnValuesSize);
 
 	/// The constructor and the fallback function doesn't to jump out.
-	if (!_function.isConstructor() && !_function.isFallback())
-		m_context.appendJump(eth::AssemblyItem::JumpType::OutOfFunction);
+	if (!_function.isConstructor())
+	{
+		solAssert(m_context.numberOfLocalVariables() == 0, "");
```
Comment (axic) [inline libsolidity/codegen/ContractCompiler.cpp:503]: Why don't we need to assert that for the constructor?
